### ISSUE TITLE
[1.13-rc5] can't get v2 Network Plugin to work

### ISSUE THREAD

**Issue by bboreham** [docker/docker#30024]
**Description**

We have a Docker Network plugin that works outside of Swarm Mode.  I tried to update it to the "v2" scheme described at https://github.com/docker/docker/blob/1.13.x/docs/extend/index.md

Docker does not seem to like what I have done, and I am unable to determine why.

**Steps to reproduce the issue:**
1. Plugin is built from the branch at https://github.com/weaveworks/weave/pull/2727
2. I pushed to Docker Hub under my personal account (bboreham) for testing
3. Results on a fresh VM below:

**Describe the results you received:**

```
vagrant@host1:~$ docker swarm init --advertise-addr=192.168.48.11
Swarm initialized: current node (f5i2dnyon5vuov1ixef022p4w) is now a manager.
[...]
vagrant@host1:~$ docker plugin install bboreham/weave2
Plugin "bboreham/weave2" is requesting the following privileges:
 - network: [host]
 - capabilities: [CAP_SYS_ADMIN CAP_NET_ADMIN]
Do you grant the above permissions? [y/N] y
latest: Pulling from bboreham/weave2
4964d439247c: Download complete 
Digest: sha256:bc4e3783a181cd14187fd62e916de2ba3a6af7e16b92ff265b6517d1c32c5b4f
Status: Downloaded newer image for bboreham/weave2:latest
Installed plugin bboreham/weave2
vagrant@host1:~$ docker plugin ls
ID                  NAME                     DESCRIPTION                   ENABLED
cca95451e51a        bboreham/weave2:latest   Weave Net plugin for Docker   true
vagrant@host1:~$ docker network create --driver=bboreham/weave2 foo
Error response from daemon: could not resolve driver bboreham/weave2 in registry
vagrant@host1:~$ docker network create --driver=bboreham/weave2:latest foo
ty438nzqwoma7bm19wa8fpvtt
vagrant@host1:~$ sudo docker-runc list
ID                                                                 PID         STATUS      BUNDLE                                                                                       CREATED
cca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5   13814       running     /run/docker/libcontainerd/cca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5   2017-01-10T10:38:47.214882347Z
vagrant@host1:~$ sudo docker-runc exec cca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5 cat /tmp/plugin.log
INFO: 2017/01/10 10:38:48.411306 Weave plugin git-a083b4f962df Command line options: [--meshsocket= --docker-api=]
INFO: 2017/01/10 10:38:48.411394 Running without Docker API connection
INFO: 2017/01/10 10:38:48.411451 Listening on /run/docker/plugins/weave.sock for global scope
INFO: 2017/01/10 10:38:50.227310 [net] GetCapabilities 
vagrant@host1:~$ docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
f244795fea11        bridge              bridge              local
edf38bb53f9b        docker_gwbridge     bridge              local
ty438nzqwoma        foo                                     swarm
08d84454e598        host                host                local
jc2zsss67mur        ingress             overlay             swarm
8b954c0d6711        none                null                local
vagrant@host1:~$ sudo journalctl -u docker
[...]
Jan 10 10:38:47 host1 docker[13260]: time="2017-01-10T10:38:47.223111091Z" level=warning msg="Unable to connect to plugin: /run/docker/plugins/cca95451e51a45602b94ba27b5f9
Jan 10 10:38:48 host1 docker[13260]: time="2017-01-10T10:38:48.223944808Z" level=warning msg="Unable to connect to plugin: /run/docker/plugins/cca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5/weave.sock/NetworkDriver.GetCapabilities: Post http://%2Frun%2Fdocker%2Fplugins%2Fcca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5%2Fweave.sock/NetworkDriver.GetCapabilities: dial unix /run/docker/plugins/cca95451e51a45602b94ba27b5f9e1f2e0f631de00f902a6a61b3b3ae8a25fa5/weave.sock: connect: no such file or directory, retrying in 1s"
Jan 10 10:38:55 host1 docker[13260]: time="2017-01-10T10:38:55.118328222Z" level=info msg="agent: found node update" module="node/agent"
Jan 10 10:39:18 host1 docker[13260]: time="2017-01-10T10:39:18.256723304Z" level=error msg="Handler for POST /v1.25/networks/create returned error: could not resolve driver bboreham/weave2 in registry"
Jan 10 10:39:26 host1 docker[13260]: time="2017-01-10T10:39:26.633462100Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 1s"
Jan 10 10:39:27 host1 docker[13260]: time="2017-01-10T10:39:27.634155742Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 2s"
Jan 10 10:39:29 host1 docker[13260]: time="2017-01-10T10:39:29.635697105Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 4s"
Jan 10 10:39:33 host1 docker[13260]: time="2017-01-10T10:39:33.636444090Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 8s"
Jan 10 10:39:41 host1 docker[13260]: time="2017-01-10T10:39:41.637361238Z" level=error msg="Failed allocation for network ty438nzqwoma7bm19wa8fpvtt" error="failed during network allocation for network ty438nzqwoma7bm19wa8fpvtt: failed while allocating driver state for network ty438nzqwoma7bm19wa8fpvtt: plugin not found"
Jan 10 10:39:41 host1 docker[13260]: time="2017-01-10T10:39:41.637899771Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 1s"
Jan 10 10:39:42 host1 docker[13260]: time="2017-01-10T10:39:42.638750103Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 2s"
Jan 10 10:39:44 host1 docker[13260]: time="2017-01-10T10:39:44.640047054Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 4s"
Jan 10 10:39:48 host1 docker[13260]: time="2017-01-10T10:39:48.641084389Z" level=warning msg="Unable to locate plugin: bboreham/weave2:latest, retrying in 8s"
```

**Describe the results you expected:**

I expected Docker to find the plugin under the name it was installed; I expected it to call `CreateNetwork` on the plugin.

**Additional information you deem important (e.g. issue happens only occasionally):**

It seems really difficult to debug a v2 plugin - no `docker logs` anything like that.

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc5
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 00:55:26 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc5
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 00:55:26 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.13.0-rc5
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bboreham/weave2:latest bridge host macvlan null overlay
Swarm: active
 NodeID: f5i2dnyon5vuov1ixef022p4w
 Is Manager: true
 ClusterID: glamg9n2s8y9lofb0ha8ucz66
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.48.11
 Manager Addresses:
  192.168.48.11:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 51371867a01c467f08af739783b8beafc154c4d7
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-51-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 362.4 MiB
Name: host1
ID: QZ5X:WYXT:D6O2:ZUXE:HYVE:4SNX:PMHJ:LY3C:4J7E:AIIT:HJID:3RR7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Test host is a VM running under VirtualBox

**Comment by justincormack** [comment]
cc @anusha-ragunathan @tiborvass help appreciated!

**Comment by mavenugo** [comment]
The missing driver name in `docker network ls` : `ty438nzqwoma        foo                                     swarm` seems incorrect. I will look into this shortly.

**Comment by mavenugo** [comment]
@tiborvass @anusha-ragunathan the daemon logs indicates that plugins-v1 is used to locate the plugin and it is failing. That is strange because, the `docker network ls` clearly spells it as a `swarm` scoped plugin and that determination is done based on `GetCapabilities` returning Global scoped and it seems atleast libnetwork's call to plugingetter package works. Is it possible that swarmkit is unable to locate the plugin ?

**Comment by mavenugo** [comment]
@bboreham unfortunately, we couldnt fix the issue in swarmkit and hence Plugin-V2 network plugins doesn't work in swarm-mode 😢
Also plugin-v1 doesnt work due to the recently introduced filtering logic and it is incompatible with the lazy-loading plugin-v1 concept. Hence, we disabled the network plugin support in 1.13.0 for swarm-mode. This is not a regression since the support wasnt available in 1.12.0 as well.

**Comment by justincormack** [comment]
reopening until merged into here.

cc @bboreham

**Comment by mavenugo** [comment]
We can close this now via https://github.com/docker/docker/pull/30145.
This fix will be ported to 1.13.1 branch shortly.
